src/github.com/matrix-org/dendrite/appservice/routing/user.go

```diff
@@ -0,0 +1,64 @@
+package routing
+
+import (
+	"encoding/json"
+	"github.com/matrix-org/dendrite/clientapi/jsonerror"
+	"github.com/matrix-org/dendrite/common/config"
+	"github.com/matrix-org/util"
+	"io/ioutil"
+	"net/http"
+)
+
+// URIToUIDResponse represents response to an AppService URI to User Id
+// (/_matrix/app/r0/user?uri={url_encoded_uri}) request
+type URIToUIDResponse struct {
+	UserID string `json:"user_id"`
+}
+
+// URIToUID implements `/_matrix/app/r0/user?uri={url_encoded_uri}`, which
+// enables users to contact App Service users directly by taking an encoded
+// URI and turning it into a Matrix ID on the homeserver.
+// https://matrix.org/docs/spec/application_service/unstable.html#user-ids
+func URIToUID(req *http.Request, cfg config.Dendrite) util.JSONResponse {
+	uri := req.URL.Query().Get("uri")
+	if uri == "" {
+		return util.JSONResponse{
+			Code: http.StatusOK,
+			JSON: nil,
+		}
+	}
+	baseReqURL := "http://" + string(cfg.Matrix.ServerName) + "/_matrix/app/unstable/thirdparty/user/"
```

Comment:
Why do you try to trigger that URL? Isn't it right that this will hit dendrite again (which does not implement this URL currently btw)?
Or which instance should answer that request?

Reply:
@anoadragon453 made some relevant comments above. Basically, I think I'm using [this endpoint](https://matrix.org/docs/spec/application_service/unstable.html#get-matrix-app-unstable-thirdparty-user-protocol) to access the third party network and get the right data back. I think I messed up, though, since I used the homeserver URL here instead of the app service URL. I'll go ahead and change that.